src/Routers/FilesRouter.js

```diff
@@ -108,87 +108,74 @@ export class FilesRouter {
+  let {
+    start, end
+  } = getRange(req);
+
+  const notEnded = (!end && end !== 0);
+  const notStarted = (!start && start !== 0);
+  // No end provided, we want all bytes
+  if (notEnded) {
+    end = stream.length - 1;
   }
-
-  if (!partialend) {
-    if (((stream.length - 1) - start) < (buffer_size)) {
-      end = stream.length - 1;
-    }else{
-      end = start + (buffer_size);
-    }
-    chunksize = (end - start) + 1;
+  // No start profided, we're reading backwards
+  if (notStarted) {
+    start = stream.length - end;
+    end = start + end - 1;
   }
 
-  if (start == 0 && end == 2) {
-    chunksize = 1;
+  // Data exceeds the buffer_size, cap
+  if (end - start >= buffer_size) {
+    end = start + buffer_size - 1;
   }
 
+  const contentLength = (end - start) + 1;
+
   res.writeHead(206, {
     'Content-Range': 'bytes ' + start + '-' + end + '/' + stream.length,
     'Accept-Ranges': 'bytes',
-    'Content-Length': chunksize,
+    'Content-Length': contentLength,
     'Content-Type': contentType,
   });
 
   stream.seek(start, function () {
     // get gridFile stream
-    var gridFileStream = stream.stream(true);
-    var bufferAvail = 0;
-    var range = (end - start) + 1;
-    var totalbyteswanted = (end - start) + 1;
-    var totalbyteswritten = 0;
+    const gridFileStream = stream.stream(true);
+    let bufferAvail = 0;
+    let remainingBytesToWrite = contentLength;
+    let totalBytesWritten = 0;
     // write to response
-    gridFileStream.on('data', function (buff) {
-      bufferAvail += buff.length;
-      //Ok check if we have enough to cover our range
-      if (bufferAvail < range) {
-        //Not enough bytes to satisfy our full range
-        if (bufferAvail > 0) {
-          //Write full buffer
-          res.write(buff);
-          totalbyteswritten += buff.length;
-          range -= buff.length;
-          bufferAvail -= buff.length;
-        }
-      } else {
-        //Enough bytes to satisfy our full range!
-        if (bufferAvail > 0) {
-          const buffer = buff.slice(0,range);
-          res.write(buffer);
-          totalbyteswritten += buffer.length;
-          bufferAvail -= range;
-        }
+    gridFileStream.on('data', function (data) {
+      bufferAvail += data.length;
+      if (bufferAvail > 0) {
+        // slice returns the same buffer if overflowing
+        // safe to call in any case
+        const buffer = data.slice(0, remainingBytesToWrite);
+        // write the buffer
+        res.write(buffer);
+        // increment total
+        totalBytesWritten += buffer.length;
+        // decrement remaining
+        remainingBytesToWrite -= data.length;
+        // decrement the avaialbe buffer
+        bufferAvail -= buffer.length;
       }
-      if (totalbyteswritten >= totalbyteswanted) {
-        //totalbytes = 0;
+      // in case of small slices, all values will be good at that point
+      // we've written enough, end...
+      if (totalBytesWritten >= contentLength) {
```

Comment:
I'm no expert at js io.  is there a try / catch / finally idiom?

maybe something like: https://stackoverflow.com/questions/4092914/java-try-catch-finally-best-practices-while-acquiring-closing-resources

Reply:
uhm, I'm not sure, I kept the original code there, as it work working as expected, the changes I put in were just because the code was a duplicate. if it throws, express will catch it :)